### PR TITLE
Initial Metrics

### PR DESCRIPTION
**Pull request by heff**
Initial example implementation of CTA metrics.

### CODE REVIEW

File: example.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+var PLAYING = 2;
+
+
+// Watch Time
+var watchTime = 0;
+var lastWatchTimeCheck;
+var watchTimeInterval;
+
+function updateWatchTime() {
+  watchTime = watchTime + (Date.now() - lastWatchTimeCheck);
```
Comment (jsanford8) [inline example.js]: Do we want to be as accurate as possible here? It's probably not worth it, but we should probably do `var now = Date.now()` at the top and use that, or someone might claim we're off by a millisecond or two

File: example.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  watchTimeInterval = window.setInterval(updateWatchTime, 250);
+});
+
+function stopWatchTimeInterval() {
+  window.clearInterval(watchTimeInterval);
+  updateWatchTime();
+}
+
+player.addEventListener('pause', stopWatchTimeInterval);
+player.addEventListener('end', stopWatchTimeInterval);
```
Comment (jsanford8) [inline example.js:36]: Time spent buffering after seek still counts in watch time, right? If so, this is good, but if not, a seek should stop it, too.

File: example.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+var watchTime = 0;
+var lastWatchTimeCheck;
+var watchTimeInterval;
+
+function updateWatchTime() {
+  watchTime = watchTime + (Date.now() - lastWatchTimeCheck);
+  lastWatchTimeCheck = Date.now();
+}
+
+player.addEventListener('playrequest', function(){
```
Comment (jsanford8) [inline example.js]: If `seek` doesn't stop the interval, you'll want to wrap `window.setInterval` in a check to see if you have one running. You'll need to do that anyway, actually, for rebuffering, or else you'll have multiple intervals.

File: example.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+player.addEventListener('pause', stopWatchTimeInterval);
+player.addEventListener('end', stopWatchTimeInterval);
+
+// Video Startup Time
+var videoStartupTime = null;
+var firstPlayRequestTime;
+
+player.addEventListener('playrequest', function(){
```
Comment (jsanford8) [inline example.js:42]: Is the ideal here to only track the _first_ play request? This will override with every subsequent `playrequest`

File: example.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  rebufferInterval = window.setInterval(updateRebufferMetrics, 250);
+});
+
+function stopRebufferInterval() {
+  window.clearInterval(rebufferInterval);
+  updateRebufferMetrics();
+}
+
+player.addEventListener('pause', stopRebufferInterval);
+player.addEventListener('playstart', stopRebufferInterval);
```
Comment (jsanford8) [inline example.js:82]: You'll want to add one of these for `seek` as well, which can interrupt a rebuffering event.

File: example.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+
+// Playback Failure
+var playbackFailure = false;
+
+player.addEventListener('error', function(event, data){
+  if (data.playbackFailure == true) {
```
Comment (jsanford8) [inline example.js:89]: Why do we care here? Can't we just remember that, so that it's either falsey or truthy?

File: example.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    playbackFailure = true;
+  }
+});
+
+
+// Exit/Abort
+var exit = false;
+
+window.addEventListener('unload', function() {
```
Comment (jsanford8) [inline example.js:98]: Why is this here/what is it for? Convincing them that ended/stopped is unnecessary, or potentially to give us a point to do soomething with all the metrics?

**Review by jsanford8 — CHANGES_REQUESTED** [review]

File: example.js (hunk 1 above)
Comment (heff) [reply example.js]: Yeah, that makes sense. With this I'm kind of aiming for something non-programmers can begin to read, but I think that fits.

File: example.js (hunk 3 above)
Comment (heff) [reply example.js]: Yeah, this assumes that you can never fire playrequest again without first going back to a paused state. Is what you're saying a safety against edge cases or something that needs to be accounted for even in expected operation?

File: example.js (hunk 2 above)
Comment (heff) [reply example.js:36]: Yeah, any buffering time counts towards watch time.

File: example.js (hunk 4 above)
Comment (heff) [reply example.js:42]: Yeah, that's a good question. Basically if it's paused before hitting playstart, is startup time from the first or second playrequest? It's kind of an edge case so in practice I could be ok either way. I'm tempted to say from the second playrequest, because otherwise we might be adding up all the playrequest times that may happen before playstart.

File: example.js (hunk 5 above)
Comment (heff) [reply example.js:82]: Damn...yeah. This is why rebuffering and seeking feel like a unique states. Whereas `error` (as long as it will be followed by a pause when it fails) does not stop rebuffering.

File: example.js (hunk 7 above)
Comment (heff) [reply example.js:98]: Just an example in case conversations come up around it. In reality we'd just send a beacon.
